### ISSUE TITLE
Null-coalescing operator cast before test against null issue

### ISSUE THREAD

**Issue by yuxiaomao** [HaxeFoundation/haxe#11706]
Here is a code which works with 4.3.4 but not with 5.0 ( https://try.haxe.org/#3F583427 ): in 5.0, using `??` cause `onEnd` first being cast to EndFun before test against null.

Thanks to Rudy I found that the test fail after this commit https://github.com/HaxeFoundation/haxe/commit/39642c94d895b73678f719f4a2fe9b5ca89bbdc8 ( #11425 )

```haxe
class Test {
	var onEnd : Void -> Void = null;
	public function new() {
	}

	public function call(f : EndFun ) {
		f();
		trace("call " + f);
	}

	static function main()
	{
		var m = new Test();
		m.call(m.onEnd == null ? function() {trace("somefunc1");} : m.onEnd); // ok
		m.call(m.onEnd ?? function() {trace("somefunc2");}); // Runtime error null access f
	}
}

@:callable abstract EndFun(Void -> Void) {
	@:from public static function fromFun( f : Void -> Void ) : EndFun {
		return cast function() {
			f();
			f = () -> throw "double end";
		};
	}
}
```

**Comment by kLabz** [comment]
(added to hotfix candidates because #11425 is in there and this issues shouldn't be missed if #11425 is to be included in 4.3.x)

**Comment by Simn** [comment]
Hmm... the problem here is that the expected type is `EndFun` because that's what `call` wants as an argument, and then we pass a `Void->Void` to it and that causes a cast. That cast should clearly happen _after_ the null-check though, not before.
